hooks/tk-multi-publish2/basic/publish_session.py

```diff
@@ -0,0 +1,733 @@
+
+        publish_path = item.properties["publish_info"]["path"]
+
+        path = item.properties["work_file_path"]
+
+        self.logger.info(
+            "Publish created for file: %s" % (publish_path,),
+            extra={
+                "action_show_in_shotgun": {
+                    "label": "Show Publish",
+                    "tooltip": "Open the Publish in Shotgun.",
+                    "entity": publish_data
+                }
+            }
+        )
+
+        # insert the path into the properties
+        item.properties["next_version_path"] = self._bump_file_version(
+            path, settings, item)
+
+    def _bump_file_version(self, path, settings, item):
+        """
+        Save the supplied path to the next version on disk.
+        """
+
+        (next_version_path, version) = self._get_next_version_info(path, item)
+
+        if version is None:
+            self.logger.debug(
+                "No version number detected in the publish path. "
+                "Skipping the bump file version step."
+            )
+            return None
+
+        self.logger.info("Incrementing session file version number...")
+
+        # nothing to do if the next version path can't be determined or if it
+        # already exists.
+        if not next_version_path:
+            self.logger.warning("Could not determine the next version path.")
+            return None
+        elif os.path.exists(next_version_path):
+            self.logger.warning(
+                "The next version of the path already exists",
+                extra={
+                    "action_show_folder": {
+                        "path": next_version_path
+                    }
+                }
+            )
+            return None
+
+        # save the session to the new path
+        _save_session(next_version_path)
+        self.logger.info("Session saved as: %s" % (next_version_path,))
+
+        return next_version_path
+
+    def _get_next_version_info(self, path, item):
+        """
+        Return the next version of the supplied path.
+
+        If templates are configured, use template logic. Otherwise, fall back to
+        the zero configuration, path_info hook logic.
+
+        :param str path: A path with a version number.
+        :param item: The session item
+
+        :return: A tuple of the form::
+
+            # the first item is the supplied path with the version bumped by 1
+            # the second item is the new version number
+            (next_version_path, version)
+        """
+
+        if not path:
+            self.logger.debug("Path is None. Can not determine version info.")
+            return None, None
+
+        publisher = self.parent
+
+        # if the session item has a known work file template, see if the path
+        # matches. if not, warn the user and provide a way to save the file to
+        # a different path
+        work_template = item.properties.get("work_file_template")
```

Comment:
here see the work template pulled form item properties as populated by the collector.

Reply:
this feels like a pretty good pattern!!!!